Add CPU Fallback

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #78972
* #78523
* __->__ #78522
* #78677

This adds the option to convert tensors to fallback to cpu if the operator is not supported on `meta` tensors. The output of the cpu fallback operator must be a new-unaliased TensorImpl because the conversion to cpu and back will lose shared metadata between inputs and outputs.

Commit:
Add CPU Fallback

[ghstack-poisoned]

Commit:
Update on "Add CPU Fallback"




[ghstack-poisoned]

Commit:
Update on "Add CPU Fallback"




[ghstack-poisoned]

Commit:
Update on "Add CPU Fallback"




[ghstack-poisoned]

eellison committed:
Update on "Add CPU Fallback"


This adds the option to convert tensors to fallback to cpu if the operator is not supported on `meta` tensors. The output of the cpu fallback operator must be a new-unaliased TensorImpl because the conversion to cpu and back will lose shared metadata between inputs and outputs.

[ghstack-poisoned]

ezyang commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -56,6 +57,17 @@ def _is_tensor_constructor(func: OpOverload):
+    try:
+        yield
+    finally:
+        cpu_fallback_enabled = orig
```
Why not have this on the fake tensor mode itself?

eellison replied:
I guess... is this something that we view as temporary? If so, it might be nice not to commit it to the FakeTensorMode api. But I guess it's not that different either way.. I can move into the Mode

ezyang replied:
It is simpler and organizationally better for it to live on the mode, so yes I'd prefer it!

Re how tensors can get to it, that would be done by storing the mode on the tensor!

ezyang replied:
Converter should have a weak ref to the input tensor

ezyang commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -169,7 +216,12 @@ def check_non_fake_tensor(x):
+        r = run_function(func, types, args, kwargs)
+    except NotImplementedError as not_implemented_error:
+        if not cpu_fallback_enabled:
+            raise not_implemented_error
```
`raise` here is better as doing it this way will destroy the original backtrace info

eellison replied:
sorry, how is this different from one i'm doing now ?

ezyang commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -87,6 +99,39 @@ def __call__(self, t, device=None):
+    with no_dispatch():
+        def to_cpu(e):
+            if isinstance(e, FakeTensor):
+                return torch.empty_like(e, device="cpu")
```
let's do `zeros_like` instead, it's safer for indexing ops

ezyang commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -87,6 +99,39 @@ def __call__(self, t, device=None):
+            if e in tensor_impls:
+                raise orig_not_implemented_exception
+
+        tree_map(throw_on_reused_impls, r)
```
This seems very insufficient and also the case being tested for here feels like it can be handled.

It's insufficient in that this fallback will not work for view operations, and you won't capture errors here because view operators will return a fresh tensor with shared storage. Testing for shared storage is better.

It's handleable in that if an output of the cpu operation is exactly the same as the input, we know we can just return the original meta in that case and it will be OK.

ezyang replied:
Yes. But you can also have an operator that doesn't return the input but does mutate the metadata and you have no way of detecting this. Well, tags would help.

eellison committed:
Update on "Add CPU Fallback"


This adds the option to convert tensors to fallback to cpu if the operator is not supported on `meta` tensors. The output of the cpu fallback operator must be a new-unaliased TensorImpl because the conversion to cpu and back will lose shared metadata between inputs and outputs.

[ghstack-poisoned]

eellison committed:
Update on "Add CPU Fallback"


This adds the option to convert tensors to fallback to cpu if the operator is not supported on `meta` tensors. The output of the cpu fallback operator must be a new-unaliased TensorImpl because the conversion to cpu and back will lose shared metadata between inputs and outputs.

[ghstack-poisoned]

eellison committed:
Update on "Add CPU Fallback"


This adds the option to convert tensors to fallback to cpu if the operator is not supported on `meta` tensors. The output of the cpu fallback operator must be a new-unaliased TensorImpl because the conversion to cpu and back will lose shared metadata between inputs and outputs.

[ghstack-poisoned]

eellison committed:
Update on "Add CPU Fallback"


This adds the option to convert tensors to fallback to cpu if the operator is not supported on `meta` tensors. The output of the cpu fallback operator must be a new-unaliased TensorImpl because the conversion to cpu and back will lose shared metadata between inputs and outputs.

[ghstack-poisoned]

eellison committed:
Update on "Add CPU Fallback"


This adds the option to convert tensors to fallback to cpu if the operator is not supported on `meta` tensors. The output of the cpu fallback operator must be a new-unaliased TensorImpl because the conversion to cpu and back will lose shared metadata between inputs and outputs.

[ghstack-poisoned]

eellison committed:
Update on "Add CPU Fallback"


This adds the option to convert tensors to fallback to cpu if the operator is not supported on `meta` tensors. The output of the cpu fallback operator must be a new-unaliased TensorImpl because the conversion to cpu and back will lose shared metadata between inputs and outputs.

[ghstack-poisoned]

eellison committed:
Update on "Add CPU Fallback"


This adds the option to convert tensors to fallback to cpu if the operator is not supported on `meta` tensors. The output of the cpu fallback operator must be a new-unaliased TensorImpl because the conversion to cpu and back will lose shared metadata between inputs and outputs.

[ghstack-poisoned]

eellison committed:
Update on "Add CPU Fallback"


This adds the option to convert tensors to fallback to cpu if the operator is not supported on `meta` tensors. The output of the cpu fallback operator must be a new-unaliased TensorImpl because the conversion to cpu and back will lose shared metadata between inputs and outputs.

[ghstack-poisoned]

eellison committed:
Update on "Add CPU Fallback"


This adds the option to convert tensors to fallback to cpu if the operator is not supported on `meta` tensors. The output of the cpu fallback operator must be a new-unaliased TensorImpl because the conversion to cpu and back will lose shared metadata between inputs and outputs.

[ghstack-poisoned]

eellison committed:
Update on "Add CPU Fallback"


This adds the option to convert tensors to fallback to cpu if the operator is not supported on `meta` tensors. The output of the cpu fallback operator must be a new-unaliased TensorImpl because the conversion to cpu and back will lose shared metadata between inputs and outputs.

[ghstack-poisoned]

eellison committed:
Update on "Add CPU Fallback"


This adds the option to convert tensors to fallback to cpu if the operator is not supported on `meta` tensors. The output of the cpu fallback operator must be a new-unaliased TensorImpl because the conversion to cpu and back will lose shared metadata between inputs and outputs.

[ghstack-poisoned]

eellison committed:
Update on "Add CPU Fallback"


This adds the option to convert tensors to fallback to cpu if the operator is not supported on `meta` tensors. The output of the cpu fallback operator must be a new-unaliased TensorImpl because the conversion to cpu and back will lose shared metadata between inputs and outputs.

[ghstack-poisoned]

eellison committed:
Update on "Add CPU Fallback"


This adds the option to convert tensors to fallback to cpu if the operator is not supported on `meta` tensors. The output of the cpu fallback operator must be a new-unaliased TensorImpl because the conversion to cpu and back will lose shared metadata between inputs and outputs.

[ghstack-poisoned]

eellison committed:
Update on "Add CPU Fallback"


This adds the option to convert tensors to fallback to cpu if the operator is not supported on `meta` tensors. The output of the cpu fallback operator must be a new-unaliased TensorImpl because the conversion to cpu and back will lose shared metadata between inputs and outputs.

[ghstack-poisoned]

eellison committed:
Update on "Add CPU Fallback"


This adds the option to convert tensors to fallback to cpu if the operator is not supported on `meta` tensors. The output of the cpu fallback operator must be a new-unaliased TensorImpl because the conversion to cpu and back will lose shared metadata between inputs and outputs.

[ghstack-poisoned]

eellison committed:
Update on "Add CPU Fallback"


This adds the option to convert tensors to fallback to cpu if the operator is not supported on `meta` tensors. The output of the cpu fallback operator must be a new-unaliased TensorImpl because the conversion to cpu and back will lose shared metadata between inputs and outputs.

[ghstack-poisoned]

eellison committed:
Update on "Add CPU Fallback"


This adds the option to convert tensors to fallback to cpu if the operator is not supported on `meta` tensors. The output of the cpu fallback operator must be a new-unaliased TensorImpl because the conversion to cpu and back will lose shared metadata between inputs and outputs.

[ghstack-poisoned]

eellison committed:
Update on "Add CPU Fallback"


This adds the option to convert tensors to fallback to cpu if the operator is not supported on `meta` tensors. The output of the cpu fallback operator must be a new-unaliased TensorImpl because the conversion to cpu and back will lose shared metadata between inputs and outputs.

[ghstack-poisoned]

eellison commented:
@pytorchbot merge
